### PR TITLE
WLT-356: heavy sequence record index

### PR DESCRIPTION
**Pull request by uhzxl**
**- What I did**
Added the ability to retrieve records from storage on HM node in an ordered form starting from a certain position
**- How I did it**
Added three DB key scopes: one to save values first and top synced pulse, one to save some pulse meta information (What is the current/next pulse? How much records?) and yet one to save record position and reference on the record itself. Made call to save that index when the HM node receives a message with records from LME.
**- How to verify it**
Run unit tests
**- Description for the changelog**
Added record sequence index on HM.

### CODE REVIEW

File: ledger/heavy/handler/handler.go [hunk 1]
```diff
@@ -43,6 +43,7 @@ type Handler struct {
 	IndexAccessor  object.IndexAccessor
 	IndexModifier  object.IndexModifier
 	DropModifier   drop.Modifier
+	recordSequence object.SequenceRecordModifier
```
Comment (egorikas) [inline ledger/heavy/handler/handler.go]: Where is being inited?

File: ledger/object/sequence.go [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+		(pos.PulseNumber == other.PulseNumber && pos.Index > other.Index)
+}
+
+type SequenceRecord struct {
+	Position SequenceRecordPosition
+	ID       insolar.ID
+	Record   record.MaterialRecord
+}
+
+type SequenceRecordAccessor interface {
```
Comment (egorikas) [inline ledger/object/sequence.go]: Add comments to the interfaces too, please

File: ledger/object/sequence.go [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+type SequenceRecordStorage interface {
+	SequenceRecordAccessor
+	SequenceRecordModifier
+	SequenceRecordCursor
+
+	// UpdatePulse makes "upsert" for top pulse number and sets at old pulse descriptor link to next pulse
+	UpdatePulse(ctx context.Context, pn insolar.PulseNumber) error
+}
+
+type SequenceRecordMemory struct{}
```
Comment (egorikas) [inline ledger/object/sequence.go]: Struct is exported. Add comments please

File: ledger/object/sequence.go [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+// ForPosition returns sequence record for provided record position.
+func (s *SequenceRecordDB) ForPosition(ctx context.Context, pos SequenceRecordPosition) (SequenceRecord, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	return s.get(ctx, pos)
+}
+
+// FromPosition returns sequence record stream from provided record position with corresponding limit.
+func (s *SequenceRecordDB) FromPosition(ctx context.Context, pos SequenceRecordPosition, limit uint32) chan interface{} {
```
Comment (egorikas) [inline ledger/object/sequence.go]: I think, It's a good idea to return empty chan of interfaces with two different values.
It'd be better to return two channels, with exact result and error

**Comment by egorikas** [comment]
And I suppose, It'd be great to write some unit-tests for the functional at least

File: ledger/object/sequence.go (hunk 4 above)
Comment (uhzxl) [reply ledger/object/sequence.go]: Okay, but I wanted to do something similar to how it works in java Reactive Streams (projectreactor.io) when data and errors occurring during the processing of a stream are transmitted through a single stream. What are the pitfalls here?

Comment (egorikas) [reply ledger/object/sequence.go]: It's not  java. Working through empty interfaces isn't a great thing in go. https://go-proverbs.github.io/

In our experience, casting of empty interfaces is an infinite root of problems

Comment (andreyromancev) [reply ledger/object/sequence.go]: Using empty interfaces is generally frowned upon in Go community. Let's consider other options.

File: ledger/heavy/handler/handler.go (hunk 1 above)
Comment (uhzxl) [reply ledger/heavy/handler/handler.go]: Added init

File: ledger/object/sequence.go (hunk 2 above)
Comment (uhzxl) [reply ledger/object/sequence.go]: Added

File: ledger/object/sequence.go (hunk 3 above)
Comment (uhzxl) [reply ledger/object/sequence.go]: Added

File: ledger/object/sequence.go (hunk 4 above)
Comment (uhzxl) [reply ledger/object/sequence.go]: Okay, I replaced it with a simple slice and error value

File: ledger/object/sequence.go [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+
+func (k topKey) Scope() store.Scope {
+	return store.ScopeReplicaPosition
+}
+
+func (k topKey) ID() []byte {
+	return []byte{byte(k)}
+}
+
+type pulseLabel struct {
```
[the remark this answers is not in the document]
Comment (andreyromancev) [reply ledger/object/sequence.go]: Actually, this struct is not required at all, if we implement iteration on DB component.

File: ledger/object/sequence.go [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+
+func (k positionKey) ID() []byte {
+	return append(SequenceRecordPosition(k).PulseNumber.Bytes(),
+		utils.UInt32ToBytes(SequenceRecordPosition(k).Index)...)
+}
+
+func (s *SequenceRecordDB) topPulse() (insolar.PulseNumber, error) {
+	key := topKey(42)
+	buff, err := s.db.Get(key)
+	if err == ErrNotFound {
```
Comment (andreyromancev) [inline ledger/object/sequence.go]: `err == store.ErrNotFound`

File: ledger/object/sequence.go [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+	return store.ScopeSequenceRecord
+}
+
+func (k positionKey) ID() []byte {
+	return append(SequenceRecordPosition(k).PulseNumber.Bytes(),
+		utils.UInt32ToBytes(SequenceRecordPosition(k).Index)...)
+}
+
+func (s *SequenceRecordDB) topPulse() (insolar.PulseNumber, error) {
+	key := topKey(42)
```
Comment (andreyromancev) [inline ledger/object/sequence.go]: Let's not commit unfinished methods.

File: ledger/object/sequence.go [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+	buff, err := s.db.Get(key)
+	if err == ErrNotFound {
+		return 0, ErrNoSequenceSyncPulses
+	} else if err != nil {
+		return 0, err
+	}
+	return insolar.PulseNumber(binary.BigEndian.Uint32(buff)), nil
```
Comment (andreyromancev) [inline ledger/object/sequence.go]: `insolar.NewPulseNumber(buff)`

File: ledger/object/sequence.go [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+	return s.get(ctx, pos)
+}
+
+// FromPosition returns sequence record slice from provided record position with corresponding limit.
+func (s *SequenceRecordDB) FromPosition(ctx context.Context, pos SequenceRecordPosition, limit uint32) ([]SequenceRecord, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	out := []SequenceRecord{}
+	capacity := uint32(0)
```
Comment (andreyromancev) [inline ledger/object/sequence.go]: This can be replaced by `len(out)`

File: ledger/object/sequence.go [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+		if err != nil {
+			return nil, errors.Wrap(err, "[SequenceRecordIndex] failed to get pulse meta")
+		}
+		for ; pos.Index <= curr.topIndex && capacity < limit; pos.Index++ {
+			rec, err := s.get(ctx, pos)
+			if err != nil {
+				return nil, errors.Wrap(err, "[SequenceRecordIndex] failed to get record")
+			}
+			out = append(out, rec)
+			capacity++
```
Comment (andreyromancev) [inline ledger/object/sequence.go]: We don't check the limit here. We can overflow it and the `capacity == limit` check will fail and iterate to the end.

File: ledger/object/sequence.go [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+	return out, nil
+}
+
+// UpdatePulse makes "upsert" for top pulse number and sets at old pulse descriptor link to next pulse
+func (s *SequenceRecordDB) UpdatePulse(ctx context.Context, pn insolar.PulseNumber) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	prev, err := s.topPulse()
+	if err == ErrNotFound {
```
Comment (andreyromancev) [inline ledger/object/sequence.go]: `ErrNoSequenceSyncPulses` is returned from `topPulse`, not `ErrNotFound`.
I propose to remove `ErrNoSequenceSyncPulses` completely and use `ErrNotFound` instead.

File: ledger/object/sequence.go [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+		}
+		return s.updateTop(pn)
+	} else if err != nil {
+		return errors.Wrap(err, "[SequenceRecordIndex] failed to get top pulse number")
+	}
+	pd, err := s.exactPulse(prev)
+	if err != nil {
+		return errors.Wrap(err, "[SequenceRecordIndex] failed to get previous pulse descriptor")
+	}
+	pd.next = pn
```
Comment (andreyromancev) [inline ledger/object/sequence.go]: `pd` is not written in db. Changes will be lost.

File: ledger/object/sequence.go (hunk 3 above)
Comment (uhzxl) [reply ledger/object/sequence.go]: Maybe for testing purposes?

**Comment by afiskon** [comment]
@uhzxl Please resolve the conflicts, make changes named during code review and re-submit PR.

**Pull request closed without merging**